R/zzz.R

```diff
@@ -42,8 +42,52 @@ methods::setOldClass("R7_object")
     stop(msg, call. = FALSE)
   }
 }
+#' @export
+`[.R7_object` <- function(x, ..., drop = TRUE) {
+  if (typeof(x) %in% c("list", "environment")) {
+    NextMethod()
+  } else {
+    msg <- "R7 object not subsettable."
+    stop(msg, call. = FALSE)
+  }
+}
 
-
+#' @export
+`[.R7_object` <- function(x, ..., drop = TRUE) {
+  if (typeof(x) %in% c("list", "environment")) {
+    NextMethod()
+  } else {
+    msg <- "R7 object not subsettable."
+    stop(msg, call. = FALSE)
+  }
+}
+#' @export
+`[<-.R7_object` <- function(x, ..., value) {
```

Comment:
interesting. I was a little surprised by this suggestion as the signature of the two methods is not the same. 

I'm also surprised this works for both cases. In the case where the error is thrown, i.e not subclass of list, I guess it works because the dispatch is dictated by the `x` parameter and since `value` isn't being used it isn't ever going to cause an issue. But in the list subclass case, this gives a method whose signature does not contain a `value` parameter explicitly before it gets delegated to `[<-.list`. 

I won't argue with your suggestion, it's your package afterall, but I would love to know why this works out ok as we effectively then have

```
`[<-.R7_object` <- function(x, ..., drop = TRUE) {
  if (typeof(x) %in% c("list", "environment")) {
    NextMethod()
  } else {
    msg <- "R7 objects are not subsettable."
    stop(msg, call. = FALSE)
  }
}

foo <- new_class("foo", class_list)
x <- foo()
x[1] <- 1
```

where I have always understood that I require the value parameter for the subset assignment operation

Reply:
Whilst the test cases for this pass subject to the change, the R CMD checks do give a warning which causes the pipeline to fail.

```
checking replacement functions ... WARNING
    ‘[<-.R7_object’ ‘[[<-.R7_object’
  The argument of a replacement function which corresponds to the right
  hand side must be named ‘value’.
```

I will revert this change in light of this, but I would still love an explanation as to why the tests pass in this case if you have one you are happy to give.

Reply:
Oh oops, I forgot that the signatures are different. In that case, I'd suggest pulling out the repeated code into a helper function. I'd make `allow_env` (or similar) as a parameter so you can use for all 4 methods.

Reply:
`check_subsettable` introduced to reduce duplication, plus associated test

Reply:
Perfect, thanks!